ci: Run "macOS native x86_64" job on GitHub Actions

hebasto opened the pull request:
From https://github.com/bitcoin/bitcoin/issues/28098:
> Thus, someone would have to sponsor an amount of roughly 5kUSD/mo for those two tasks.

> If the goal is to stay on a free plan, I think the only option is GitHub Actions CI.

---

**IMPORTANT NOTE**. We currently ship macOS release binaries for both architectures: `x86_64` and `arm64`. If this PR gets merged, only `x86_64` architecture will be tested on CI, which implies some [drawbacks](https://github.com/bitcoin/bitcoin/pull/28187#issuecomment-1658077549).

However, it has never been the case that our CI tested both architectures simultaneously. And we hope that GitHub Actions will soon host macOS `arm64` runners.

Historically, we moved from `x86_64` to `arm64` in https://github.com/bitcoin/bitcoin/pull/26388 less than a year ago.

---

Security concerns:
- https://github.com/bitcoin/bitcoin/issues/28098#issuecomment-1651432106
- https://github.com/bitcoin/bitcoin/issues/28098#issuecomment-1651688197

`GITHUB_TOKEN` permissions (from the build log in my personal repo):
```
2023-07-27T07:30:17.8313534Z ##[group]GITHUB_TOKEN Permissions
2023-07-27T07:30:17.8314113Z Contents: read
2023-07-27T07:30:17.8314608Z Metadata: read
2023-07-27T07:30:17.8314957Z Packages: read
2023-07-27T07:30:17.8315233Z ##[endgroup]
```

Comparison of resources:

| Resource | Current, Cirrus CI | Suggested, GitHub Actions |
|---|:-:|:-:|
| CPU | 4 | 4 \*\* |
| RAM, GB | 8 | 14 |

**\*\* NOTE**: However, [docs](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners#supported-runners-and-hardware-resources) are mentioning:
> 3-core CPU (x86_64)

fanquake commented:
This is replacing the arm64 job with an x86_64 job?

hebasto commented:
> This is replacing the arm64 job with an x86_64 job?

Yes. No `arm64` macOS images are available in GitHub Actions for now.

fanquake commented:
The PR description should probably mention that, as it's a regression in terms of testing. i.e no-longer testing on Apples primarily supported hardware.

hebasto commented:
> The PR description should probably mention that, as it's a regression in terms of testing. i.e no-longer testing on Apples primarily supported hardware.

Done.

maflcko reviewed (approved):
lgtm, Should squash?

fanquake commented:
~0 Is there any reason we can't use the public beta of the arm64 runners? Have we tested that at-least?

maflcko reviewed (approved):
> ~0 Is there any reason we can't use the public beta of the arm64 runners? Have we tested that at-least?

Can you add a link to documentation on how to set that up? With details on the runners, etc. In any case, I don't think it matters, unless you can point to a bug that was found with arm64, but not with amd64. In practice, I expect all bugs found by CI to be found by both macOS arches.

fanquake commented:
> Can you add a link to documentation on how to set that up?

Trying to find that at the moment. Maybe it doesn't exist.

> unless you can point to a bug that was found with arm64, but not with amd64.

I think it's at least worth bringing up, as the PR as presented, didn't mention that it drops (direct) testing of one of our release platforms entirely, and presented this change as-if just swapping underlying CI provider.

In terms of differences, there are code paths that are currently exercised in our CI, that no-longer would be after this change, i.e use of macOS ARM intrinsics, so we may be less-likely to detect any breakage if related code is changed, i.e can't sanity check in the CI that detection is working. Maybe that doesn't matter, I do agree that most bugs should be found in either case.

hebasto commented:
> Can you link to 10 passing tests in your fork? Otherwise we may run into intermittent issues.

1. https://github.com/hebasto/bitcoin/actions/runs/5708399487/attempts/1
2. https://github.com/hebasto/bitcoin/actions/runs/5708399487/attempts/2
3. https://github.com/hebasto/bitcoin/actions/runs/5708399487/attempts/3
4. https://github.com/hebasto/bitcoin/actions/runs/5708399487/attempts/4
5. https://github.com/hebasto/bitcoin/actions/runs/5708399487/attempts/5
6. https://github.com/hebasto/bitcoin/actions/runs/5708399487/attempts/6
7. https://github.com/hebasto/bitcoin/actions/runs/5708399487/attempts/7
8. https://github.com/hebasto/bitcoin/actions/runs/5708399487/attempts/8
9. https://github.com/hebasto/bitcoin/actions/runs/5708399487/attempts/9
10. https://github.com/hebasto/bitcoin/actions/runs/5708399487/attempts/10

All other comments are to be addressed shortly.

hebasto commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,45 @@
+        uses: actions/cache@v3
+        with:
+          path: /Users/runner/work/bitcoin/bitcoin/ci/scratch/.ccache
+          key: ${{ runner.os }}-${{ runner.arch }}-ccache-cache-${{ github.run_id }}
+          restore-keys: ${{ runner.os }}-${{ runner.arch }}-ccache-cache
```
> any reason to not just use the task name, like before?

GHA's job (an analogue of Cirrus's task) [context](https://docs.github.com/en/actions/learn-github-actions/contexts#job-context) has no "name"-like property.

hebasto replied:
> Ok. You are setting one in the line above: `name: macOS 13 native, x86_64 [no depends, sqlite only, gui]`
> 
> But if that is not possible to use, that is fine.

Well, it is still possible to get `github.job` property that has value `macos-native` from https://github.com/bitcoin/bitcoin/blob/e9b72185205812df51082b42fb8c25cbf53cbc03/.github/workflows/ci.yml#L14-L15

> Though, it would be good to explain why both `key` and `restore-key` are set and why `github.run_id` is used?

This use case is described in `actions/cache` [docs](https://github.com/actions/cache/blob/main/tips-and-workarounds.md#update-a-cache). TL;DR: it forces to update the cache regardless of its successful restoring.

hebasto replied:
[Updated](https://github.com/bitcoin/bitcoin/pull/28187#issuecomment-1665619844).

hebasto commented:
> > All other comments are to be addressed shortly.
> 
> Are you still working on this?

I am. My apologies for a delay.

hebasto commented on .cirrus.yml:
```diff
@@ -323,20 +323,3 @@ task:
-    << : *CIRRUS_EPHEMERAL_WORKER_TEMPLATE_ENV
-    CI_USE_APT_INSTALL: "no"
-    PACKAGE_MANAGER_INSTALL: "echo"  # Nothing to do
-    FILE_ENV: "./ci/test/00_setup_env_mac_native_arm64.sh"
```
I hope that at some point, GitHub Actions will host `arm64` macOS runners, enabling us to test both architectures.

Maybe leave both files for now?

hebasto replied:
[Updated](https://github.com/bitcoin/bitcoin/pull/28187#issuecomment-1665619844).

hebasto commented:
Updated e9b72185205812df51082b42fb8c25cbf53cbc03 -> cc26f0d8c8539094db8d95da9fa46ea1247a0d8c ([pr28187.03](https://github.com/hebasto/bitcoin/commits/pr28187.03) -> [pr28187.04](https://github.com/hebasto/bitcoin/commits/pr28187.04)):

- rebased
- squashed
- addressed @MarcoFalke's comments:
  - https://github.com/bitcoin/bitcoin/pull/28187#discussion_r1279021074
  - https://github.com/bitcoin/bitcoin/pull/28187#discussion_r1278964021
  - https://github.com/bitcoin/bitcoin/pull/28187#discussion_r1278968101

hebasto commented on .github/workflows/ci.yml:
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/28187#issuecomment-1665619844).

hebasto commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,45 @@
+      - name: Ccache cache
+        uses: actions/cache@v3
+        with:
+          path: /Users/runner/work/bitcoin/bitcoin/ci/scratch/.ccache
```
[Reworked](https://github.com/bitcoin/bitcoin/pull/28187#issuecomment-1665619844).

hebasto commented:
From [IRC](https://bitcoin-irc.chaincode.com/bitcoin-core-dev/2023-08-03#952553):
> 14:22 \<fanquake\> Yea the PR description in that PR needs updating to actually explain what’s its doing (dropping support for a release platform from CI), rather than making it look like just swapping infra

The PR description has been updated.

maflcko reviewed (approved):
lgtm ACK cc26f0d8c8539094db8d95da9fa46ea1247a0d8c 🖐

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: lgtm ACK cc26f0d8c8539094db8d95da9fa46ea1247a0d8c 🖐
hvJ/DOYbi+VSxM5yYqoUvgK/U86A7rpQTGv2wNi1lLHZOushQXVRrU9nznuLgMptqRzWzBwp4zMzw3g0Vz/1Aw==
```

</details>

hebasto commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,51 @@
+    name: macOS 13 native, x86_64 [no depends, sqlite only, gui]
+    # Use latest image, but hardcode version to avoid silent upgrades (and breaks).
+    # See: https://github.com/actions/runner-images#available-images.
+    runs-on: macos-13
```
[[Usage limits](https://docs.github.com/en/actions/learn-github-actions/usage-limits-billing-and-administration#usage-limits)](https://docs.github.com/en/actions/learn-github-actions/usage-limits-billing-and-administration#usage-limits):
> Each job in a workflow can run for up to 6 hours of execution time.

hebasto replied:
Done.

hebasto commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,51 @@
+        run: clang --version
+
+      - name: Install Homebrew packages
+        run: brew install automake boost pkg-config miniupnpc libnatpmp zeromq qt@5 qrencode ccache gnu-getopt
```
Restored.

jarolrod commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,53 @@
+name: bitcoin-core-ci
+run-name: Bitcoin Core CI
+# See: https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#onpushpull_requestpull_request_targetpathspaths-ignore.
+on: [pull_request, push]
```
I'm catching up on the syntax for Github Actions, but wouldn't this run this CI task twice when a Pull request is opened? That seems to be the case when opening this against my fork: https://github.com/jarolrod/bitcoin/pull/2

hebasto replied:
> That seems to be the case when opening this against my fork: [jarolrod#2](https://github.com/jarolrod/bitcoin/pull/2)

That's correct. But is this really the case we want to care about?

hebasto committed:
ci: Run "macOS native x86_64" job on GitHub Actions

Also, the "macOS native arm64" task has been removed from Cirrus CI.

hebasto commented:
Updated fe34609476b11cdd907c298f7300d424bf556e98 -> 9658d0dc17c270138523c41a982425e276b24271 ([pr28187.05](https://github.com/hebasto/bitcoin/commits/pr28187.05) -> [pr28187.06](https://github.com/hebasto/bitcoin/commits/pr28187.06), [diff](https://github.com/hebasto/bitcoin/compare/pr28187.05..pr28187.06)):

- addressed @MarcoFalke's [comment](https://github.com/bitcoin/bitcoin/pull/28187#discussion_r1288122926)
- the workflow run name is set as the [commit message](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#run-name)
- push events for tags are skipped

hebasto commented:
> Anything left to do here, for a required CI-only pull request with review?

I believe this PR is ready to merge unless @fanquake has more comment regarding its description.

However, it requires adjusting "Actions permissions" and "Workflow permissions" in the repo in a similar way as it was done in https://github.com/bitcoin-core/secp256k1/pull/1389.

@achow101 Would you mind taking care of that, please?

jarolrod commented on .cirrus.yml:
```diff
@@ -334,20 +334,3 @@ task:
     << : *CIRRUS_EPHEMERAL_WORKER_TEMPLATE_ENV
-
-task:
-  name: 'macOS 13 native arm64 [gui, sqlite only] [no depends]'
```
nit: if there is any opinion on the pr/commit title not being sufficient, you could make it something like the following (and add a body): 

```
ci: Run "macOS native" job on Github Actions, revert back to x86_64

Updates the macOS native ci job to run on Github Actions which will
allow us to stay on a free plan for the job. Additionally, temporarily
revert back to running on a x86_64 architecture until Github Actions
has a arm64 (Apple Silicon) runner; which is preferable to no macOS native
ci job.
```

jarolrod reviewed (commented):
ACK 9658d0dc17c270138523c41a982425e276b24271

As long as permissions are looked over and activation of Github Workflows and Actions is safe here: https://docs.github.com/en/rest/actions/permissions?apiVersion=2022-11-28

achow101 commented:
ACK 9658d0dc17c270138523c41a982425e276b24271

fanquake commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,59 @@
+        with:
+          path: ${{ env.CCACHE_DIR }}
+          key: ${{ github.job }}-ccache-cache-${{ github.run_id }}
+          restore-keys: ${{ github.job }}-ccache-cache
```
Yea, looks like caching doesn't work at all here?

hebasto replied:
> Yea, looks like caching doesn't work at all here?

- https://github.com/bitcoin/bitcoin/issues/21552
- https://github.com/bitcoin/bitcoin/pull/24620

fanquake replied:
If it doesn't work, why was any of this added to the CI template?

hebasto replied:
> If it doesn't work, why was any of this added to the CI template?

Ccache regression, introduced in https://github.com/bitcoin/bitcoin/pull/19683, should be fixed. It is not related to the CI template that is correct.

hebasto replied:
> How is this related. This is the "macOS native", not "macOS-cross" task.

My bad. Sorry for the noise.

hebasto replied:
> Yea, looks like caching doesn't work at all here?

It [works](https://github.com/bitcoin/bitcoin/actions/runs/5878420889/job/15940465042).

> I think it works when there are no pull requests. It should be possible to fix this in a follow-up.

Yes, pull requests should not create their own caches. Going to address this issue shortly.

hebasto commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,59 @@
+    # No need to run on the read-only mirror, unless it is a PR.
+    if: github.repository != 'bitcoin-core/gui' || github.event_name == 'pull_request'
+
+    timeout-minutes: 120
```
GHActions have no such a feature by default. It should be a separated workflow that processes [`pull_request.synchronize`](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#pull_request) event and terminates other workflows. However, I'm not sure about the exact implementation for now.

fanquake commented:
> Looks like it is now even harder to re-run tasks,

So there is no way for developers to re-run there own PRs at all, with GitHub Actions? Other than I assume force pushing?

hebasto commented:
> > Looks like it is now even harder to re-run tasks,
> 
> So there is no way for developers to re-run there own PRs at all, with GitHub Actions? Other than I assume force pushing?

Docs are not super clear about that: https://docs.github.com/en/actions/managing-workflow-runs/re-running-workflows-and-jobs.

fanquake commented:
> That page says "People with write permissions to a repository can re-run workflows in the repository." 

Hopefully not. If this really is the case, that's another downside to GH Actions, and going to be a wrose developer experience (we can't give write access to all contributors just to re-run CI). Would have been good if these things had been explained in the PR description, or discussion.

> can confirm whether they can re-run on Cirrus?

Contributors can definitely re-run Cirrus jobs. Looks like they cannot re-run Github Actions workflows.

hebasto commented:
> > Maybe someone else without write access can confirm whether they can re-run on Cirrus?
> 
> I can't re-run GH actions tasks but am able to re-run cirrus jobs.

Does closing a PR and following re-opening it work?

dergoegge commented:
> Does closing a PR and following re-opening it work?

No, that just cancels all jobs and the gha one just continues running

hebasto commented:
> Also, it looks like CI tasks currently need approval from a maintainer to run at all from non-members. See [docs.github.com/en/actions/managing-workflow-runs/approving-workflow-runs-from-public-forks](https://docs.github.com/en/actions/managing-workflow-runs/approving-workflow-runs-from-public-forks)

It is a configurable setting.
